lib/qbo_api/connection/oauth2.rb

```diff
@@ -10,15 +54,47 @@ def self.included(*)
 
     def default_attributes
       super.merge!(
-        access_token: nil
+        access_token:   nil,
+        refresh_token:  nil,
+        client_id:      defined?(CLIENT_ID) ? CLIENT_ID : nil,
+        client_secret:  defined?(CLIENT_SECRET) ? CLIENT_SECRET : nil,
+        refresh_proc:   nil
       )
     end
+
     def add_oauth2_authorization_middleware(conn)
       conn.request :oauth2, access_token, token_type: 'bearer'
+      conn.use FaradayMiddleware::OAuth2Refresh, self if refresh_token
     end
 
     def use_oauth2_middleware?
-      access_token != nil
+      refresh_token || access_token
+    end
+  end
+end
+
+
+# @private
+module FaradayMiddleware
+  # @private
+  class OAuth2Refresh < Faraday::Middleware
+    AUTH_HEADER = 'Authorization'.freeze
+
+    def call(env)
+      begin
+        @app.call(env).tap do |resp|
+          raise QboApi::Unauthorized if resp.status == 401
+        end
+      rescue QboApi::Unauthorized => _error
+        env[:request_headers][AUTH_HEADER] = "Bearer #{@qbo_api.refresh_access_token!}"
+        retry
```

Comment:
we'll also want to handle invalid grant etc

it's a good practice to limit retries, so 👎 

```ruby
def call(env, retries=1)
   resp = @app.call(env)
   case resp.status
   when 401
      if retries > 0
       retries -= 1
       env[:request_headers][AUTH_HEADER] = "Bearer #{@qbo_api.refresh_access_token!}"
       retry
     else
       QboApi::Unauthorized, "some useful info"
      end
   else resp
   end
````